rfcs/20190802-model-garden-redesign.md

```diff
@@ -0,0 +1,207 @@
+
+| Status        | Proposed       |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Jing Li (jingli@google.com), Hongkun Yu (hongkuny@google.com), Xiaodan Song (xiaodansong@google.com) |
+| **Sponsor**   | Edd Wilder-James (ewj@google.com)                 |
+| **Updated**   | 2019-08-02                                           |
+
+## Objective
+
+This document presents a proposal to redesign TensorFlow official model garden.
+We aim to provide a central and reliable place to contain popular examples,
+state-of-the-art models and tutorials to demonstrate the best practice in TF2.0
+and illustrate real-world use cases.
+
+## Motivation
+
+The current [TF official model garden](https://github.com/tensorflow/models/tree/master/official)
+mainly has ad hoc support. Example models are implemented using mixed TensorFlow
+APIs in different coding styles and some of them have convergence and/or
+performance regression. With TensorFlow 2.0 launch, there’s a great desire to
+provide tensorflow users a clear and central place to showcase reliable TF2.0
+models with the best practices to follow.
+
+We want to take this opportunity to substantially improve the state of the
+official model garden, and provide seamlessly end-to-end training and inference
+user experience on a wide range of accelerators and mobile device chips. We hope
+to encourage community to contribute innovations and improve TensorFlow
+efficiency and usability.
+
+## User Benefit
+
+We aim to provide the best modeling experience via this revamp effort:
+
+*   Usability and reliability
+    *   keep official models well-maintained and tested for both performance and
+        convergence.
+    *   provide accessible model distribution via [TensorFlow Hub](https://www.tensorflow.org/hub) and share state-of-the-art research accomplishments.
+    *   make training on both GPU and TPU an easy switch.
+    *   provide reusable components for research and production.
+*   End-to-end solutions
+    *   provide seamless end-to-end training and inference solutions, where inference covers serving on TPU, GPU, mobile and edge devices.
+    *   provide hyper parameter sets to tune models for various resource constraints.
+    *   provide solutions with hyper parameters to scale model training to TPU pods or multi-worker GPUs.
+    *   provide variants derived from standard models to tackle various practical tasks.
+
+## Design Proposal
+
+### Official model directory reorgnization
+
+We are going to reorganize the official model directory to provide:
+
+*   common libraries, mainly two types:
+    *   Common training util library in TF2.0, model configuration and
+        hyperparameter definition in a consistent style.
+    *   Model category related common library, e.g. primitives as basic building
+        block for NLP models. We will follow the fundamental design of Keras
+        layer/network/model to define and utilize model building blocks.
```

Comment:
Note: `Network` is a private class, so objects will need to subclass either `Layer` or `Model`.

Also note that whenever possible, we should prefer the pattern of Functional model + custom layers, rather than subclassed models (which have less functionality).

Reply:
Thanks for the suggestions! We do plan to follow the pattern of functional model + custom layers.